src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java

```diff
@@ -0,0 +1,158 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.command;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import net.minecraft.command.CommandException;
+import net.minecraft.command.EntitySelector;
+import net.minecraft.command.ICommandSender;
+import net.minecraft.entity.Entity;
+import net.minecraftforge.fml.common.FMLLog;
+import net.minecraftforge.fml.common.Loader;
+
+/**
+ * Allows registration of custom selector types by assigning a {@link SelectorHandler} to a prefix
+ * To check whether a new prefix conflicts with any registered prefix, the methods {@link #isShadowed} and {@link #isShadowing} can be used
+ * This class handles calls to the {@link EntitySelector} methods {@link EntitySelector#matchEntities matchEntities},
+ * {@link EntitySelector#matchesMultiplePlayers matchesMultiplePlayers} and {@link EntitySelector#hasArguments hasArguments}.<br>
+ * The calls are delegated to the handler with the longest matching prefix
+ */
+public class SelectorHandlerManager
+{
+    private SelectorHandlerManager()
+    {
+    }
+
+    //the ordering is reversed such that longer prefixes appear before their shorter substrings
+    private static final SortedMap<String, SelectorHandler> selectorHandlers = new TreeMap<String, SelectorHandler>(Collections.<String> reverseOrder());
+    private static final SortedMap<String, String> registeringMods = new TreeMap<String, String>(Collections.<String> reverseOrder());
+
+    private static final SelectorHandler vanillaHandler = new SelectorHandler()
+    {
+        @Override
+        public <T extends Entity> List<T> matchEntities(final ICommandSender sender, final String token, final Class<? extends T> targetClass) throws CommandException
+        {
+            return EntitySelector.matchEntitiesDefault(sender, token, targetClass);
+        }
+
+        @Override
+        public boolean matchesMultiplePlayers(final String selectorStr) throws CommandException
+        {
+            return EntitySelector.matchesMultiplePlayersDefault(selectorStr);
+        }
+
+        @Override
+        public boolean isSelector(final String selectorStr)
+        {
+            return EntitySelector.isSelectorDefault(selectorStr);
+        }
+    };
+
+    static
+    {
+        for (final String prefix : ArrayUtils.toArray("@p", "@a", "@r", "@e"))
+        {
+            selectorHandlers.put(prefix, vanillaHandler);
+            registeringMods.put(prefix, "Minecraft");
+        }
+    }
+
+    /**
+     * Returns whether there is already a more specific prefix registered
+     */
+    public static boolean isShadowed(final String prefix)
+    {
+        for (final String other : registeringMods.keySet())
```

Comment:
Unfortunately, I don't think so:
Assume you have prefixes `a` and `aa` registered. When you call this method with `ac`, the current code would return `true` (since `"ac".startsWith("a")`). `floorKey` however returns `null` (when the order is reversed) and `ceilingKey` returns `aa`, which is also not what we want.

`headMap`/`tailMap` (as you suggested for `getHandler`) or maybe even `subMap` should work however - we'll look into it and update the PR accordingly.